File: app/src/main/java/com/dt/manager/widget/CodeEditorView.java
package com.dt.manager.widget;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.text.Editable;
import android.text.Layout;
import android.text.Spannable;
import android.text.TextWatcher;
import android.text.style.CharacterStyle;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.ScaleGestureDetector;
import android.view.ViewConfiguration;

import androidx.appcompat.widget.AppCompatEditText;

import com.dt.manager.core.SyntaxHighlighter;

/**
 * EditText that paints a line-number gutter on the left side and
 * (optionally) applies syntax highlighting. Supports pinch-to-zoom
 * to change the text size.
 *
 * The view should be wrapped inside a NestedScrollView (vertical)
 * + HorizontalScrollView (horizontal) for proper scroll behavior.
 * Set layout_width and layout_height to wrap_content.
 */
public class CodeEditorView extends AppCompatEditText {

    private static final int GUTTER_WIDTH_DP = 36;
    private static final int GUTTER_PADDING_DP = 6;
    private static final float DEFAULT_TEXT_SIZE_SP = 11f;
    private static final float MIN_TEXT_SIZE_SP = 7f;
    private static final float MAX_TEXT_SIZE_SP = 28f;

    private Paint gutterBgPaint;
    private Paint lineNumberPaint;
    private Paint separatorPaint;
    private int gutterWidthPx;
    private int gutterPaddingPx;

    private float textSizeSp = DEFAULT_TEXT_SIZE_SP;
    private ScaleGestureDetector scaleDetector;

    private SyntaxHighlighter.Language language = SyntaxHighlighter.Language.TEXT;
    private boolean highlightEnabled = true;
    private boolean suppressHighlight = false;

    private final Runnable highlightRunnable = this::applyHighlight;

    public CodeEditorView(Context context) {
        super(context);
        init();
    }

    public CodeEditorView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public CodeEditorView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        float density = getContext().getResources().getDisplayMetrics().density;
        gutterWidthPx = (int) (GUTTER_WIDTH_DP * density);
        gutterPaddingPx = (int) (GUTTER_PADDING_DP * density);

        gutterBgPaint = new Paint();
        gutterBgPaint.setColor(0xFF2B2B2B);
        gutterBgPaint.setStyle(Paint.Style.FILL);

        separatorPaint = new Paint();
        separatorPaint.setColor(0xFF3A3A3A);
        separatorPaint.setStyle(Paint.Style.FILL);

        lineNumberPaint = new Paint();
        lineNumberPaint.setColor(0xFF808080);
        lineNumberPaint.setTypeface(Typeface.MONOSPACE);
        lineNumberPaint.setAntiAlias(true);
        lineNumberPaint.setTextSize(DEFAULT_TEXT_SIZE_SP * density);
        lineNumberPaint.setTextAlign(Paint.Align.RIGHT);

        setTypeface(Typeface.MONOSPACE);
        setTextColor(0xFFEEEEEE);
        setBackgroundColor(Color.TRANSPARENT);
        setTextSize(DEFAULT_TEXT_SIZE_SP);
        setLineSpacing(3 * density, 1.0f);
        setHorizontallyScrolling(true);
        setGravity(Gravity.TOP);

        setPadding(gutterWidthPx + gutterPaddingPx,
                (int) (8 * density),
                (int) (12 * density),
                (int) (8 * density));

        // Pinch-to-zoom
        scaleDetector = new ScaleGestureDetector(getContext(), new ScaleGestureDetector.SimpleOnScaleGestureListener() {
            @Override
            public boolean onScale(ScaleGestureDetector detector) {
                float factor = detector.getScaleFactor();
                float newSize = textSizeSp * factor;
                if (newSize < MIN_TEXT_SIZE_SP) newSize = MIN_TEXT_SIZE_SP;
                if (newSize > MAX_TEXT_SIZE_SP) newSize = MAX_TEXT_SIZE_SP;
                if (Math.abs(newSize - textSizeSp) < 0.1f) return true;
                textSizeSp = newSize;
                setTextSize(textSizeSp);
                float density = getContext().getResources().getDisplayMetrics().density;
                lineNumberPaint.setTextSize(textSizeSp * density);
                invalidate();
                return true;
            }
        });

        addTextChangedListener(new TextWatcher() {
            @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            @Override public void onTextChanged(CharSequence s, int start, int before, int count) {}
            @Override public void afterTextChanged(Editable s) {
                if (suppressHighlight || !highlightEnabled) return;
                removeCallbacks(highlightRunnable);
                postDelayed(highlightRunnable, 250);
            }
        });
    }

    public void setLanguage(SyntaxHighlighter.Language language) {
        this.language = language;
        if (highlightEnabled) applyHighlight();
    }

    public void setHighlightEnabled(boolean enabled) {
        this.highlightEnabled = enabled;
        if (enabled) applyHighlight();
    }

    public void setTextSizeSp(float sp) {
        textSizeSp = sp;
        if (textSizeSp < MIN_TEXT_SIZE_SP) textSizeSp = MIN_TEXT_SIZE_SP;
        if (textSizeSp > MAX_TEXT_SIZE_SP) textSizeSp = MAX_TEXT_SIZE_SP;
        setTextSize(textSizeSp);
        float density = getContext().getResources().getDisplayMetrics().density;
        lineNumberPaint.setTextSize(textSizeSp * density);
        invalidate();
    }

    public float getTextSizeSp() {
        return textSizeSp;
    }

    /** Apply syntax highlighting to the current text. */
    private void applyHighlight() {
        Editable e = getText();
        if (e == null || e.length() == 0) return;

        // Clear existing spans
        CharacterStyle[] spans = e.getSpans(0, e.length(), CharacterStyle.class);
        for (CharacterStyle s : spans) {
            e.removeSpan(s);
        }

        if (language == SyntaxHighlighter.Language.TEXT) return;

        Spannable highlighted = SyntaxHighlighter.highlight(e.toString(), language);
        CharacterStyle[] newSpans = highlighted.getSpans(0, highlighted.length(), CharacterStyle.class);
        suppressHighlight = true;
        for (CharacterStyle s : newSpans) {
            int start = highlighted.getSpanStart(s);
            int end = highlighted.getSpanEnd(s);
            e.setSpan(CharacterStyle.wrap(s), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
        }
        suppressHighlight = false;
    }

    @Override
    public boolean onTouchEvent(android.view.MotionEvent event) {
        scaleDetector.onTouchEvent(event);
        if (scaleDetector.isInProgress()) {
            return true;
        }
        return super.onTouchEvent(event);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        // Draw gutter background
        canvas.drawRect(0, 0, gutterWidthPx, getHeight(), gutterBgPaint);

        Layout layout = getLayout();
        if (layout != null) {
            int lineCount = getLineCount();
            int firstVisibleLine = layout.getLineForVertical(getScrollY());
            int lastVisibleLine = layout.getLineForVertical(getScrollY() + getHeight());
            if (lastVisibleLine >= lineCount - 1) lastVisibleLine = lineCount - 1;
            if (lastVisibleLine < 0) lastVisibleLine = 0;

            for (int i = firstVisibleLine; i <= lastVisibleLine; i++) {
                int baseline = layout.getLineBaseline(i) + getPaddingTop();
                String numText = String.valueOf(i + 1);
                float x = gutterWidthPx - gutterPaddingPx;
                canvas.drawText(numText, x, baseline, lineNumberPaint);
            }
        }

        // Vertical separator between gutter and code
        canvas.drawRect(gutterWidthPx, 0, gutterWidthPx + 1, getHeight(), separatorPaint);

        super.onDraw(canvas);
    }
}

File: app/src/main/java/com/dt/manager/widget/CodeEditorView.kt
package com.dt.manager.widget

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.text.Editable
import android.text.Spannable
import android.text.TextWatcher
import android.text.style.CharacterStyle
import android.util.AttributeSet
import android.view.Gravity
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import androidx.appcompat.widget.AppCompatEditText
import com.dt.manager.core.SyntaxHighlighter
import kotlin.math.abs

/**
 * EditText that paints a line-number gutter on the left side and
 * applies syntax highlighting. Supports pinch-to-zoom to change the text size.
 */
class CodeEditorView : AppCompatEditText {

    companion object {
        private const val GUTTER_WIDTH_DP = 36
        private const val GUTTER_PADDING_DP = 6
        private const val DEFAULT_TEXT_SIZE_SP = 11f
        private const val MIN_TEXT_SIZE_SP = 7f
        private const val MAX_TEXT_SIZE_SP = 28f
    }

    private lateinit var gutterBgPaint: Paint
    private lateinit var lineNumberPaint: Paint
    private lateinit var separatorPaint: Paint
    private var gutterWidthPx = 0
    private var gutterPaddingPx = 0

    var textSizeSp = DEFAULT_TEXT_SIZE_SP
        private set
    private lateinit var scaleDetector: ScaleGestureDetector

    var language: SyntaxHighlighter.Language = SyntaxHighlighter.Language.TEXT
        set(value) {
            field = value
            if (highlightEnabled) applyHighlight()
        }

    var highlightEnabled = true
        set(value) {
            field = value
            if (value) applyHighlight()
        }

    private var suppressHighlight = false
    private val highlightRunnable = Runnable { applyHighlight() }

    constructor(context: Context) : super(context) { init() }
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) { init() }
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) { init() }

    private fun init() {
        val density = context.resources.displayMetrics.density
        gutterWidthPx = (GUTTER_WIDTH_DP * density).toInt()
        gutterPaddingPx = (GUTTER_PADDING_DP * density).toInt()

        gutterBgPaint = Paint().apply {
            color = -0xd4d4d5 // 0xFF2B2B2B
            style = Paint.Style.FILL
        }

        separatorPaint = Paint().apply {
            color = -0xc5c5c6 // 0xFF3A3A3A
            style = Paint.Style.FILL
        }

        lineNumberPaint = Paint().apply {
            color = -0x7f7f80 // 0xFF808080
            typeface = Typeface.MONOSPACE
            isAntiAlias = true
            textSize = DEFAULT_TEXT_SIZE_SP * density
            textAlign = Paint.Align.RIGHT
        }

        typeface = Typeface.MONOSPACE
        setTextColor(-0x111112) // 0xFFEEEEEE
        setBackgroundColor(Color.TRANSPARENT)
        setTextSize(DEFAULT_TEXT_SIZE_SP)
        setLineSpacing(3 * density, 1.0f)
        setHorizontallyScrolling(true)
        gravity = Gravity.TOP

        setPadding(
            gutterWidthPx + gutterPaddingPx,
            (8 * density).toInt(),
            (12 * density).toInt(),
            (8 * density).toInt()
        )

        scaleDetector = ScaleGestureDetector(context, object : ScaleGestureDetector.SimpleOnScaleGestureListener() {
            override fun onScale(detector: ScaleGestureDetector): Boolean {
                val factor = detector.scaleFactor
                val newSize = (textSizeSp * factor).coerceIn(MIN_TEXT_SIZE_SP, MAX_TEXT_SIZE_SP)
                if (abs(newSize - textSizeSp) < 0.1f) return true
                textSizeSp = newSize
                setTextSize(textSizeSp)
                val d = context.resources.displayMetrics.density
                lineNumberPaint.textSize = textSizeSp * d
                invalidate()
                return true
            }
        })

        addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                if (suppressHighlight || !highlightEnabled) return
                removeCallbacks(highlightRunnable)
                postDelayed(highlightRunnable, 250)
            }
        })
    }

    fun setTextSizeSp(sp: Float) {
        textSizeSp = sp.coerceIn(MIN_TEXT_SIZE_SP, MAX_TEXT_SIZE_SP)
        setTextSize(textSizeSp)
        val density = context.resources.displayMetrics.density
        lineNumberPaint.textSize = textSizeSp * density
        invalidate()
    }

    private fun applyHighlight() {
        val e = text ?: return
        if (e.isEmpty()) return

        val spans = e.getSpans(0, e.length, CharacterStyle::class.java)
        for (s in spans) {
            e.removeSpan(s)
        }

        if (language == SyntaxHighlighter.Language.TEXT) return

        val highlighted = SyntaxHighlighter.highlight(e.toString(), language)
        val newSpans = highlighted.getSpans(0, highlighted.length, CharacterStyle::class.java)
        suppressHighlight = true
        for (s in newSpans) {
            val start = highlighted.getSpanStart(s)
            val end = highlighted.getSpanEnd(s)
            e.setSpan(CharacterStyle.wrap(s), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
        }
        suppressHighlight = false
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        scaleDetector.onTouchEvent(event)
        if (scaleDetector.isInProgress) {
            return true
        }
        return super.onTouchEvent(event)
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawRect(0f, 0f, gutterWidthPx.toFloat(), height.toFloat(), gutterBgPaint)

        val l = layout
        if (l != null) {
            val lineCount = lineCount
            var firstVisibleLine = l.getLineForVertical(scrollY)
            var lastVisibleLine = l.getLineForVertical(scrollY + height)
            if (lastVisibleLine >= lineCount - 1) lastVisibleLine = lineCount - 1
            if (lastVisibleLine < 0) lastVisibleLine = 0
            if (firstVisibleLine < 0) firstVisibleLine = 0

            for (i in firstVisibleLine..lastVisibleLine) {
                val baseline = l.getLineBaseline(i) + paddingTop
                val numText = (i + 1).toString()
                val x = (gutterWidthPx - gutterPaddingPx).toFloat()
                canvas.drawText(numText, x, baseline.toFloat(), lineNumberPaint)
            }
        }

        canvas.drawRect(gutterWidthPx.toFloat(), 0f, (gutterWidthPx + 1).toFloat(), height.toFloat(), separatorPaint)
        super.onDraw(canvas)
    }
}
